frigate/output/preview.py

```diff
@@ -0,0 +1,195 @@
+        playlist.append(
+            f"file '{os.path.join(CACHE_DIR, f'{FOLDER_PREVIEW_FRAMES}/preview_{self.camera}-{self.frame_times[-1]}.jpg')}'"
+        )
+
+        p = sp.run(
+            self.ffmpeg_cmd.split(" "),
+            input="\n".join(playlist),
+            encoding="ascii",
+            capture_output=True,
+        )
+
+        start = self.frame_times[0]
+        end = self.frame_times[-1]
+
+        if p.returncode == 0:
+            logger.info("successfully saved preview")
+            self.inter_process_queue.put_nowait(
+                (
+                    INSERT_PREVIEW,
+                    {
+                        Previews.id: f"{end}-{start}",
+                        Previews.camera: self.camera,
+                        Previews.path: self.path,
+                        Previews.start_time: start,
+                        Previews.end_time: end,
+                        Previews.duration: end - start,
+                    },
+                )
+            )
+        else:
+            logger.error(f"Error saving preview for {self.camera} :: {p.stderr}")
+
+        # unlink files from cache
+        for t in self.frame_times:
+            Path(
+                os.path.join(
+                    CACHE_DIR, f"{FOLDER_PREVIEW_FRAMES}/preview_{self.camera}-{t}.jpg"
+                )
+            ).unlink(missing_ok=True)
+
+
+class PreviewRecorder:
+    def __init__(self, config: CameraConfig, inter_process_queue: mp.Queue) -> None:
+        self.config = config
+        self.inter_process_queue = inter_process_queue
+        self.start_time = 0
+        self.last_output_time = 0
+        self.output_frames = []
+        self.out_height = 160
+        self.out_width = int(
+            (config.detect.width / config.detect.height) * self.out_height
+        )
+
+        Path(os.path.join(CACHE_DIR, "preview_frames")).mkdir(exist_ok=True)
+        Path(os.path.join(CLIPS_DIR, f"previews/{config.name}")).mkdir(
+            parents=True, exist_ok=True
+        )
+
+    def should_write_frame(
+        self,
+        current_tracked_objects: list[dict[str, any]],
+        motion_boxes: list[list[int]],
+        frame_time: float,
+    ) -> bool:
+        """Decide if this frame should be added to summary."""
+        # limit output to 1 fps
+        if (frame_time - self.last_output_time) < 1 / SUMMARY_OUTPUT_FPS:
+            return False
+
+        # send frame if a non-stationary object is in a zone
+        if any(
+            (len(o["current_zones"]) > 0 and not o["stationary"])
+            for o in current_tracked_objects
+        ):
+            self.last_output_time = frame_time
+            return True
+
+        # TODO think of real motion box logic to use
+        if len(motion_boxes) % 2 == 1:
+            self.last_output_time = frame_time
+            return True
+
+        return False
+
+    def write_frame_to_cache(self, frame_time: float, frame) -> None:
+        frame = cv2.cvtColor(
```

Comment:
Is it possible resize the yuv frame and use that to avoid the jpg conversion overhead?

Reply:
from what I can tell it doesn't seem there is an easy way to save to .yuv file and have ffmpeg read from as concat format